packages/babel-plugin-react-jsx/__tests__/__snapshots__/TransformJSXToReactJSX-test.js.snap

```diff
@@ -38,6 +38,167 @@ var div = React.jsx(Component, Object.assign({}, props, {
 }));
 `;
 
+exports[`transform react to jsx auto import can specify source 1`] = `
+import * as _foobar from "foobar";
+
+var x = _foobar.jsx("div", {
+  children: _foobar.jsx("span", {})
+});
+`;
+
+exports[`transform react to jsx auto import default 1`] = `
+import _default from "react";
+
+var x = _default.jsx(_default.Fragment, {
+  children: _default.jsxs("div", {
+    children: [_default.jsx("div", {}, "1"), _default.jsx("div", {
+      meow: "wolf"
+    }, "2"), _default.jsx("div", {}, "3"), _default.createElement("div", Object.assign({}, props, {
+      key: "4"
+    }))]
+  })
+});
+`;
+
+exports[`transform react to jsx auto import in dev 1`] = `
+import { createElement as _createElement } from "react";
+import { jsxDEV as _jsxDEV } from "react";
+import { Fragment as _Fragment } from "react";
+var _jsxFileName = "";
+
+var x = _jsxDEV(_Fragment, {
+  children: _jsxDEV("div", {
+    children: [_jsxDEV("div", {}, "1", false, {
+      fileName: _jsxFileName,
+      lineNumber: 4
+    }, this), _jsxDEV("div", {
+      meow: "wolf"
+    }, "2", false, {
+      fileName: _jsxFileName,
+      lineNumber: 5
+    }, this), _jsxDEV("div", {}, "3", false, {
+      fileName: _jsxFileName,
+      lineNumber: 6
+    }, this), _createElement("div", Object.assign({}, props, {
+      key: "4",
+      __source: {
+        fileName: _jsxFileName,
+        lineNumber: 7
+      },
+      __self: this
+    }))]
+  }, undefined, true, {
+    fileName: _jsxFileName,
+    lineNumber: 3
+  }, this)
+}, undefined, false);
+`;
+
+exports[`transform react to jsx auto import named exports 1`] = `
+import { createElement as _createElement } from "react";
+import { jsx as _jsx } from "react";
+import { jsxs as _jsxs } from "react";
+import { Fragment as _Fragment } from "react";
+
+var x = _jsx(_Fragment, {
+  children: _jsxs("div", {
+    children: [_jsx("div", {}, "1"), _jsx("div", {
+      meow: "wolf"
+    }, "2"), _jsx("div", {}, "3"), _createElement("div", Object.assign({}, props, {
+      key: "4"
+    }))]
+  })
+});
+`;
+
+exports[`transform react to jsx auto import namespace 1`] = `
+import * as _react from "react";
+
+var x = _react.jsx(_react.Fragment, {
```

Comment:
We could add an additional option to emit a variable to cache these at the top of the file.

```
import * as _react from "react";
const _fragment = _react.Fragment;
```
...
```
var x = _jsx(_Fragment, {
```

For a native ES module system this wouldn't matter but the way most module transpilers are implemented this becomes an object. Which may need an indirection to load the object and then the current function in the object for each call. Instead of just loading the current function.

Note that this also applies to `namedExports`. It should be something like:

```
import {jsx as __jsx, Fragment as __Fragment} from "react";
const _jsx = __jsx;
const _fragment = __Fragment;
```

Modern browser VMs may optimize for the inline cache containing this information perhaps. So I don't think this matters but it might for Hermes.

Additionally this approach would slow down initialization of the module scope and use more memory.

So I doubt this approach will actually be better overall but it would be nice to have an option so we can experiment with it and test which one is faster in practice.

Reply:
Ditto for shadowing question and tests on these variables if we add this!

Reply:
I believe I tested this once with React.createElement in V8 and found it didn't make a significant difference (IIRC slightly worse; definitely wasn't better or I would've gone through with it).

Reply:
Yea I tested it as well and came to the same conclusion. So we’re not doing this.

The key is that as long as the VM optimizes for storing methods on the object hidden class it’s guaranteed to be worse once you use more than one method.

What I’m wondering next is if it would be worth while having React be a global property.